### PR TITLE
Optimize agent upsert query operations

### PR DESCRIPTION
**Pull request by ankilosaurus**
Problem
---
Entries in agents table gets updated on every ping even when there is no change. This insert/update happens in ping flow from each host Teletraan manages and is a high volume query specifically during fleet wide deploy. 

Fix
---
Don't update agents table if there is no update. This should reduce database upsert operations into agents table by 100% when there is no deploy(no state change). Note, Last_update field will not be updated on every ping.

Tests:
----
1. Deploy new build
2. Restart current build
3. Pause/resume a deploy in stage
4. Pause/resume deploy on a given host
5. Rollback deploy

### CODE REVIEW

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java [hunk 1]
```diff
@@ -277,6 +277,25 @@ AgentState proposeNewAgentState(PingReportBean report, AgentBean agent) {
     // This is intended to be used for update agent record, not for goal,
     // We populate all the fields, since this could be used for insertOrUpdate as well
     AgentBean genUpdateBeanByReport(PingReportBean report, AgentBean agent) {
+        AgentState newAgentState = proposeNewAgentState(report, agent);
+
+        // Don't update if not changed.
+        if (agent != null && agent.getHost_id() == host_id && agent.getDeploy_id() == report.getDeployId() &&
```
Comment (disparter) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:283]: Is this checks are fine? I mean, they are Strings, why not use equals instead?

**Review by disparter — CHANGES_REQUESTED** [review]
The building is failing also, after dealing with comment I think its good to go

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java (hunk 1 above)
Comment (ankilosaurus) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:283]: yup, this won't work. Working on changing it.

**Comment by ankilosaurus** [comment]
> The building is failing also, after dealing with comment I think its good to go

yeah I was expecting test failures and was surprised when it didn't fail earlier. Now it's failing on tests which is good.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java [hunk 2]
```diff
@@ -314,7 +333,7 @@ AgentBean genUpdateBeanByReport(PingReportBean report, AgentBean agent) {
     // Generate new agent bean based on the report & current agent record,
     // This is intended to be used for deploy goal to install next stage
     AgentBean genNextStageUpdateBean(EnvironBean env, PingReportBean report, AgentBean agent) {
-        AgentBean agentBean = genUpdateBeanByReport(report, agent);
+        AgentBean agentBean = genAgentBeanByReport(report, agent);
```
Comment (yongwen) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:336]: can we still call genUpdateBeanByReport() to minimize/isolate the changes. as the logic is to done mainly  inside the genUpdateBeanByReport()

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java [hunk 3]
```diff
@@ -461,7 +480,7 @@ void process(String envId, EnvironBean env, PingReportBean report, AgentBean age
         AgentBean updateBean = null;
         if (report != null) {
             updateBean = genUpdateBeanByReport(report, agent);
-            if (!StringUtils.isEmpty(report.getEnvId())) {
+            if (!StringUtils.isEmpty(report.getEnvId()) && updateBean != null) {
```
Comment (yongwen) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:483]: we are changing the logic here a lot, wondering if we can minimize the logic change and only handle not to update the DB part.  The unit test is failing so worth to try to isolate the change.

Comment (ankilosaurus) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:483]: yes, agree. I will upload new diff.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java (hunk 2 above)
Comment (ankilosaurus) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:336]: updating a diff.

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java [hunk 4]
```diff
@@ -273,6 +273,28 @@ AgentState proposeNewAgentState(PingReportBean report, AgentBean agent) {
...
+                    origBean.getEnv_id(), origBean.getDeploy_id(), origBean.getHost_id());
+            return false;
+        }
+        LOG.info("Agent record for env_id {}, deploy_id {} on host {} needs update",
+                origBean.getEnv_id(), origBean.getDeploy_id(), origBean.getHost_id());
+        return true;
+    }
+
     // Generate new agent bean based on the report & current agent record,
     // This is intended to be used for update agent record, not for goal,
```
Comment (robbintt) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:299]: update comment

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java [hunk 5]
```diff
@@ -646,6 +671,7 @@ void process(String envId, EnvironBean env, PingReportBean report, AgentBean age
             LOG.debug("GoalAnalyst case 3.1 - add an uninstall candidate to instruct host {} to remove the retired env {}", host, report.getEnvId());
             uninstallCandidates.add(new UninstallCandidate(report));
             updateBean.setState(AgentState.DELETE);
+            needUpdateAgents.put(report.getEnvId(), updateBean);
```
Comment (robbintt) [inline deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:674]: would be better if this happened implicitly when updateBean.setState()

Comment (robbintt) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java:674]: discussed: move to caller

**Review by robbintt — CHANGES_REQUESTED** [review]
see inline comments

**Review by yongwen — APPROVED** [review]

**Review by robbintt — APPROVED** [review]

**Pull request merged**